README.md

```diff
@@ -303,6 +306,42 @@ However, you should note that this option doesn't mean anything if you didn't
 persist your database volume. Because if it is not persisted, then it will be lost 
 on restart because docker will recreate the container.
 
+## Postgres SSL setup
+
+By default the image is delivered with an unsigned SSL certificate. This helps to have an
+encrypted connection to clients and avoid eavesdropping but does not help to mitigate
+man in the middle (MITM) attacks.
+
+You need to provide your own, signed private key to avoid this kind of attacks (and make
+sure clients connect with verify-ca or verify-full sslmode).
+
+The following is an example Dockerfile that sets up a container with custom ssl private key and certificate:
+
+```
+FROM kartoza/postgis:11.0-2.5
+
+ADD ssl_cert.pem /etc/ssl/certs/ssl_cert.pem
+ADD localhost_ssl_key.pem /etc/ssl/private/ssl_key.pem
+
+RUN chmod 400 /etc/ssl/private/ssl_key.pem
+```
```

Comment:
@m-kuhn how are we going to use these env variables in the context of building the image with them. Are we expecting users to use them as mounted volumes for each certificate? Assuming that the user will mount them how are we going to ensure the permissions are applied correctly

Reply:
Thanks for getting back to me @NyakudyaA .

In my scenario I went with the Dockerfile as listed here (no mounted volumes) because it's for unit testing and the certs and keys come from a git repo with random user and permissions with which I didn't want to mess on the host. But I could imagine that on a live system one could also handle it with a mounted volume with proper user/permission outside.

Does this answer the question? Or do you have a better proposal to deal with this?

Reply:
Thanks @m-kuhn I am happy with your explanation but is it always going to be a two-step process all the time according to your Dockerfile and docker-compose. If so can we not just combine the logic in your Dockerfile into the main one here https://github.com/kartoza/docker-postgis/blob/develop/Dockerfile.

Reply:
Thanks @NyakudyaA 

* What could possibly be done is removing the environment variable management and have a predefined place somewhere where custom certs need to be placed and as soon as they are found there the server will just prefer them over the snakeoil certs.
* It would be nice to be able to provide certs via volumes, but that has not been verified yet
* The Dockefile approach is not very nice, but it has been verified, so I documented that one. Integration upstream is IMO not possible because the dockerfile you linked to is used to build the image upstream and the certificates and keys for this use case are provided by the client.

Reply:
@m-kuhn I will merge this and you can use this in QGIS. But once 

> Thanks @NyakudyaA
> 
>     * What could possibly be done is removing the environment variable management and have a predefined place somewhere where custom certs need to be placed and as soon as they are found there the server will just prefer them over the snakeoil certs.

Will investigate this solution for now.
> 
>     * It would be nice to be able to provide certs via volumes, but that has not been verified yet

I will try this out and revert to you at some point
> 
>     * The Dockefile approach is not very nice, but it has been verified, so I documented that one. Integration upstream is IMO not possible because the dockerfile you linked to is used to build the image upstream and the certificates and keys for this use case are provided by the client.

Thanks for this, will merge this

Reply:
Thanks a lot.
Just ping me whenever you have more questions.
And it would be nice if you could let me know when the new image is released.

Reply:
After reading this thread, my opinions are:

1. Regarding environment variables

Keeping the environment variables is still nice to do and allow more customizations. We just need to make sure the default location is set to the preferred way


2. Regarding how to provide the secrets

Reference [1] states that we can provide the secrets using `docker secret` store. I think it is easier to setup compared with volumes. But bind mounting volumes to host path is also a possible choice. But I imagine when setting up unittests, it will be easier to define the secrets via `docker secrets` then provide the location via docker compose configuration file, just as the [1] shown in the sample.

3. Regarding "verification"

> The Dockefile approach is not very nice, but it has been verified, so I documented that one. Integration upstream is IMO not possible because the dockerfile you linked to is used to build the image upstream and the certificates and keys for this use case are provided by the client.

I don't understand what does "verified" means in this context. Is this *verified to work* as in *tested to work*.? Or *verified as the standard security best practice* (in addition of also being tested to be working for your use case)? 

But overall, I agree that if it's possible that downstream user can use the image without rebuilding it (just able to configure it), then we should use that approach while conforming to standard security best practice.


# Sample of docker compose file to configure secrets

If we use docker compose file to configure where the secret is located, we could add this sample to the README, instead of the Dockerfile sample above:

```
version: '3.1'

services:
   db:
     image: kartoza/postgis:latest
     volumes:
       - db_data:/var/lib/postgresql
     environment:
       SSL_CERT_FILE: /run/secrets/ssl_cert
       SSL_KEY_FILE: /run/secrets/ssl_key
       SSL_CA_FILE: /run/secrets/ssl_ca
     secrets:
       - source: ssl_cert
         target: ssl_cert
       - source: ssl_key
         target: ssl_key
         mode: 0400
       - source: ssl_ca
         target: ssl_ca


secrets:
   ssl_cert:
     file: ./host/location/ssl_cert.pem
   ssl_key:
     file: ./host/location/ssl_key.key
   ssl_ca:
     file: ./host/location/ssl_ca.pem
```

Reply:
Note that in the compose file above, it is recommended to set the file mode permission (400 for ssl_key) directly in the compose files, as shown in [3]

Reply:
I forgot to ping @m-kuhn for his feedback :)